dashboard/lib/certificate_image.rb

```diff
@@ -161,25 +161,125 @@ def self.create_course_certificate_image(name, course = nil, donor_name = nil, c
-      apply_text(image, total_hours.to_s, 30, 'Times bold', 'rgb(87,87,87)', -248, 124, 80, 30)
-      donor_text_y_offset = 611
-    else # all other courses use a certificate image where the course name is also blank
+    else
+      unit_or_unit_group = CurriculumHelper.find_matching_unit_or_unit_group(course)
       image = Magick::Image.read(path).first
-      apply_text(image, name, 75, 'Helvetica bold', 'rgb(118,101,160)', 0, -135, CERT_NAME_AREA_WIDTH, CERT_NAME_AREA_HEIGHT)
-      # The area in pixels which will display the course title.
-      course_title_width = 1000
-      course_title_height = 60
-      apply_text(image, course_title, 47, 'Helvetica bold', 'rgb(29,173,186)', 0, 15, course_title_width, course_title_height)
-      donor_text_y_offset = 447
+
+      text_constants = {
+        'self_paced_pl_certificate.png': {
+          name: {
+            font_size: 62,
+            x_offset: 0,
+            y_offset: -248,
+            height: 70,
+            width: CERT_NAME_AREA_WIDTH,
+          },
+          two_titles: {
+            unit_group_height: 85,
+            unit_group_font_size: 62,
+            unit_group_x_offset: 0,
+            unit_group_y_offset: -57,
+            unit_height: 71,
+            unit_font_size: 57,
+            unit_x_offset: 0,
+            unit_y_offset: 36,
+          },
+          one_title: {
+            font_size: 62,
+            x_offset: 0,
+            y_offset: 0,
+            height: 171,
+          },
+          course_title_width: 1400,
+          donor_text_y_offset: 611,
+        },
+        'blank_certificate.png': {
+          name: {
+            font_size: 75,
+            x_offset: 0,
+            y_offset: -135,
+            height: CERT_NAME_AREA_HEIGHT,
+            width: CERT_NAME_AREA_WIDTH,
+          },
+          two_titles: {
+            unit_group_width: 1000,
+            unit_group_height: 50,
+            unit_group_font_size: 40,
+            unit_group_x_offset: 0,
+            unit_group_y_offset: 0,
+            unit_width: 800,
+            unit_height: 40,
+            unit_font_size: 32,
+            unit_x_offset: 0,
+            unit_y_offset: 47,
+          },
+          one_title: {
+            font_size: 47,
+            x_offset: 0,
+            y_offset: 15,
+            height: 60,
+            width: 1000,
+          },
+          donor_text_y_offset: 447,
+        },
+      }
+      cert_text_constants = text_constants[template_file.to_sym]
+
+      name_constants = cert_text_constants[:name]
+      apply_text(image, name, name_constants[:font_size], 'Helvetica bold', 'rgb(118,101,160)', name_constants[:x_offset], name_constants[:y_offset], name_constants[:width], name_constants[:height])
+
+      # When we have a unit within a unit_group, we want to display both the unit and unit_group titles.
+      # When we have a standalone unit or the unit group, we only display the localized title of unit_or_unit_group.
```

Comment:
Thank you for refactoring this! It's easier for me to follow now🎉

Reply:
Yay! Thank you for the great review 😀